### PR TITLE
Improve decimal aggregations

### PR DESCRIPTION
**Pull request by martint**
Remove unnecessary calls to get state, which can be expensive due to
lookups in the underlying BigArrays.

https://jmh.morethan.io/?sources=https://gist.githubusercontent.com/martint/f2685c5b8fc19f8feaa2a3ea03c43bda/raw/82b40a4efcbae857f901e9f51105902790cbf036/before.json,https://gist.githubusercontent.com/martint/f2685c5b8fc19f8feaa2a3ea03c43bda/raw/82b40a4efcbae857f901e9f51105902790cbf036/after.json

### CODE REVIEW

File: core/trino-main/src/main/java/io/trino/operator/aggregation/DecimalAverageAggregation.java [hunk 1]
```diff
@@ -161,27 +161,34 @@ public static void inputLongDecimal(Type type, LongDecimalWithOverflowAndLongSta
...
-        if (state.getLongDecimal() == null) {
-            state.setLongDecimal(UnscaledDecimal128Arithmetic.unscaledDecimal());
+        Slice sum = state.getLongDecimal();
+        if (sum == null) {
+            sum = UnscaledDecimal128Arithmetic.unscaledDecimal();
+            state.setLongDecimal(sum);
         }
+        long overflow = UnscaledDecimal128Arithmetic.addWithOverflow(sum, unscaledValue, sum);
+        if (overflow != 0) {
+            state.setOverflow(state.getOverflow() + overflow);
```
Comment (pettyjamesm) [inline core/trino-main/src/main/java/io/trino/operator/aggregation/DecimalAverageAggregation.java:171]: Seems like `state.addOverflow(overflow)` would suffice here.

File: core/trino-main/src/main/java/io/trino/operator/aggregation/DecimalAverageAggregation.java [hunk 2]
```diff
@@ -161,27 +161,34 @@ public static void inputLongDecimal(Type type, LongDecimalWithOverflowAndLongSta
...
-        accumulateAndUpdateOverflow(unscaledValue, state);
-        state.setLong(state.getLong() + 1);
-    }
-
-    private static void initializeIfNeeded(LongDecimalWithOverflowAndLongState state)
-    {
-        if (state.getLongDecimal() == null) {
-            state.setLongDecimal(UnscaledDecimal128Arithmetic.unscaledDecimal());
+        Slice sum = state.getLongDecimal();
+        if (sum == null) {
```
Comment (pettyjamesm) [inline core/trino-main/src/main/java/io/trino/operator/aggregation/DecimalAverageAggregation.java:165]: See the below comment about maybe having and using something like:
 ```
 Slice sum = state.getOrInitializeLongDecimal(unscaledValue);
 if (sum != unscaledValue) {
   state.addOverflow(UnscaledDecimal128Arithmetic.addWithOverflow(sum, unscaledValue, sum));
 }
 ```

File: core/trino-main/src/main/java/io/trino/operator/aggregation/DecimalAverageAggregation.java (hunk 1 above)
Comment (martint) [reply core/trino-main/src/main/java/io/trino/operator/aggregation/DecimalAverageAggregation.java:171]: oops... yeah, missed this one

File: core/trino-main/src/main/java/io/trino/operator/aggregation/DecimalAverageAggregation.java (hunk 2 above)
Comment (pettyjamesm) [reply core/trino-main/src/main/java/io/trino/operator/aggregation/DecimalAverageAggregation.java:165]: Disregard, this strategy is ok for short decimal where we know it was just heap allocated but extremely dangerous for long decimals where you might have a Slice pointing into the middle of the input block.

File: core/trino-main/src/main/java/io/trino/operator/aggregation/DecimalAverageAggregation.java [hunk 3]
```diff
@@ -161,27 +161,34 @@ public static void inputLongDecimal(Type type, LongDecimalWithOverflowAndLongSta
...
             state.setLongDecimal(otherState.getLongDecimal());
         }
         else {
-            accumulateAndUpdateOverflow(otherState.getLongDecimal(), state);
+            overflow += UnscaledDecimal128Arithmetic.addWithOverflow(sum, otherState.getLongDecimal(), sum);
+        }
+
+        if (overflow != 0) {
+            state.addOverflow(overflow);
```
[the remark this answers is not in the document]
Comment (pettyjamesm) [reply core/trino-main/src/main/java/io/trino/operator/aggregation/DecimalAverageAggregation.java]: Disregard, this (appears) to be material in my local benchmarking. Doing the check inside of the add method happens after bimorphic method dispatch and is more expensive than localizing the check AFAICT.

File: core/trino-main/src/main/java/io/trino/operator/aggregation/DecimalAverageAggregation.java [hunk 4]
```diff
@@ -161,27 +161,34 @@ public static void inputLongDecimal(Type type, LongDecimalWithOverflowAndLongSta
...
     public static void combine(LongDecimalWithOverflowAndLongState state, LongDecimalWithOverflowAndLongState otherState)
     {
-        state.setLong(state.getLong() + otherState.getLong());
-        state.setOverflow(state.getOverflow() + otherState.getOverflow());
+        state.addLong(otherState.getLong());
 
-        if (state.getLongDecimal() == null) {
+        long overflow = otherState.getOverflow();
+
+        Slice sum = state.getLongDecimal();
```
[the remark this answers is not in the document]
Comment (martint) [reply core/trino-main/src/main/java/io/trino/operator/aggregation/DecimalAverageAggregation.java:180]: I'm thinking of another follow up change with a dedicated implementation of the aggregates for long vs short decimals, where we avoid using Slice entirely for short decimals.

File: core/trino-main/src/test/java/io/trino/operator/aggregation/BenchmarkDecimalAggregation.java [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+import static io.trino.spi.type.DecimalType.createDecimalType;
+import static io.trino.sql.analyzer.TypeSignatureProvider.fromTypes;
+
+@State(Scope.Thread)
+@OutputTimeUnit(TimeUnit.NANOSECONDS)
+@Fork(2)
+@Warmup(iterations = 10)
+@Measurement(iterations = 10)
+@BenchmarkMode(Mode.AverageTime)
+public class BenchmarkDecimalAggregation
```
Comment (skrzypo987) [inline core/trino-main/src/test/java/io/trino/operator/aggregation/BenchmarkDecimalAggregation.java:58]: You can move that to a separate commit and post before/after results in this commit message

File: core/trino-main/src/main/java/io/trino/operator/aggregation/DecimalAverageAggregation.java [hunk 6]
```diff
@@ -211,19 +214,11 @@ public static BigDecimal average(LongDecimalWithOverflowAndLongState state, Deci
         BigDecimal sum = new BigDecimal(Decimals.decodeUnscaledValue(state.getLongDecimal()), type.getScale());
         BigDecimal count = BigDecimal.valueOf(state.getLong());
 
-        if (state.getOverflow() != 0) {
+        long overflow = state.getOverflow();
+        if (overflow != 0) {
             BigInteger overflowMultiplier = TWO.shiftLeft(UNSCALED_DECIMAL_128_SLICE_LENGTH * 8 - 2);
-            BigInteger overflow = overflowMultiplier.multiply(BigInteger.valueOf(state.getOverflow()));
-            sum = sum.add(new BigDecimal(overflow));
+            sum = sum.add(new BigDecimal(overflowMultiplier.multiply(BigInteger.valueOf(overflow))));
```
Comment (pettyjamesm) [inline core/trino-main/src/main/java/io/trino/operator/aggregation/DecimalAverageAggregation.java:220]: Overflow multiplier is a constant

**Review by pettyjamesm — APPROVED** [review]

**Comment by sopel39** [comment]
[Uploading Benchmarks comparison-aggrdec.pdf…]()

File: core/trino-main/src/main/java/io/trino/operator/aggregation/DecimalAverageAggregation.java [hunk 7]
```diff
@@ -161,28 +161,31 @@ public static void inputLongDecimal(Type type, LongDecimalWithOverflowAndLongSta
...
+        long overflow = UnscaledDecimal128Arithmetic.addWithOverflow(sum, unscaledValue, sum);
+        state.addOverflow(overflow);
+        state.addLong(1);
     }
 
     public static void combine(LongDecimalWithOverflowAndLongState state, LongDecimalWithOverflowAndLongState otherState)
     {
-        state.setLong(state.getLong() + otherState.getLong());
-        state.setOverflow(state.getOverflow() + otherState.getOverflow());
+        state.addLong(otherState.getLong());
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/operator/aggregation/DecimalAverageAggregation.java:177]: it would be nice to have some comments what `addLong` means (e.g updating row counter)

File: core/trino-main/src/main/java/io/trino/operator/aggregation/DecimalAverageAggregation.java [hunk 8]
```diff
@@ -161,28 +161,31 @@ public static void inputLongDecimal(Type type, LongDecimalWithOverflowAndLongSta
...
-    }
-
-    private static void initializeIfNeeded(LongDecimalWithOverflowAndLongState state)
-    {
-        if (state.getLongDecimal() == null) {
-            state.setLongDecimal(UnscaledDecimal128Arithmetic.unscaledDecimal());
+        Slice sum = state.getLongDecimal();
+        if (sum == null) {
+            sum = UnscaledDecimal128Arithmetic.unscaledDecimal();
+            state.setLongDecimal(sum);
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/operator/aggregation/DecimalAverageAggregation.java:168]: you could skip `addWithOverflow` completely here as in `combine` method

File: core/trino-main/src/main/java/io/trino/operator/aggregation/DecimalAverageAggregation.java [hunk 9]
```diff
@@ -161,28 +161,31 @@ public static void inputLongDecimal(Type type, LongDecimalWithOverflowAndLongSta
...
         }
+        long overflow = UnscaledDecimal128Arithmetic.addWithOverflow(sum, unscaledValue, sum);
+        state.addOverflow(overflow);
+        state.addLong(1);
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/operator/aggregation/DecimalAverageAggregation.java:172]: you could move this before `if`

File: core/trino-main/src/main/java/io/trino/operator/aggregation/DecimalSumAggregation.java [hunk 10]
```diff
@@ -140,38 +140,35 @@ private static InternalAggregationFunction generateAggregation(Type inputType, T
 
     public static void inputShortDecimal(Type type, LongDecimalWithOverflowState state, Block block, int position)
     {
-        accumulateValueInState(unscaledDecimal(type.getLong(block, position)), state);
+        Slice sum = state.getLongDecimal();
+        if (sum == null) {
+            sum = UnscaledDecimal128Arithmetic.unscaledDecimal();
+            state.setLongDecimal(sum);
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/operator/aggregation/DecimalSumAggregation.java:146]: you could skip `addWithOverflow` in this branch as in `combine`

File: core/trino-main/src/main/java/io/trino/operator/aggregation/state/LongDecimalWithOverflowStateFactory.java [hunk 11]
```diff
@@ -133,6 +141,12 @@ public void setOverflow(long overflow)
             this.overflow = overflow;
         }
 
+        @Override
+        public void addOverflow(long overflow)
+        {
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/operator/aggregation/state/LongDecimalWithOverflowStateFactory.java:146]: why no `if (overflow != 0) {` check?

**Review by sopel39 — COMMENTED** [review]
small comments

File: core/trino-main/src/main/java/io/trino/operator/aggregation/state/LongDecimalWithOverflowStateFactory.java (hunk 11 above)
Comment (martint) [reply core/trino-main/src/main/java/io/trino/operator/aggregation/state/LongDecimalWithOverflowStateFactory.java:146]: Because the addition is cheap. In the grouped case, we guard it because the cause to read/set the value in the underlying array is more expensive

**Comment by martint** [comment]
Updated

File: core/trino-main/src/main/java/io/trino/operator/aggregation/DecimalAverageAggregation.java (hunk 8 above)
Comment (martint) [reply core/trino-main/src/main/java/io/trino/operator/aggregation/DecimalAverageAggregation.java:170]: Unfortunately, no. The way this aggregation manages state is a little messy. It relies on side effects to update the state: 
* it sets the state to an empty slice if it's null
* if relies on addWithOverflow to update it in place

I think there are opportunities to clean this up and improve it later, but it will require a rewrite and specialization for short vs long decimals

Comment (sopel39) [reply core/trino-main/src/main/java/io/trino/operator/aggregation/DecimalAverageAggregation.java:170]: You could use `state.setLongDecimal(UnscaledDecimal128Arithmetic#unscaledDecimal(unscaledValue))` (it does copy slice)
Are we afraid that original `unscaledValue` will be modified?

File: core/trino-main/src/main/java/io/trino/operator/aggregation/DecimalAverageAggregation.java [hunk 12]
```diff
@@ -161,28 +161,33 @@ public static void inputLongDecimal(Type type, LongDecimalWithOverflowAndLongSta
...
-        state.setLong(state.getLong() + otherState.getLong());
-        state.setOverflow(state.getOverflow() + otherState.getOverflow());
+        state.addLong(otherState.getLong()); // row counter
+
+        long overflow = otherState.getOverflow();
 
-        if (state.getLongDecimal() == null) {
+        Slice sum = state.getLongDecimal();
+        if (sum == null) {
             state.setLongDecimal(otherState.getLongDecimal());
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/operator/aggregation/DecimalAverageAggregation.java:185]: now updates in place will also affect `otherState.getLongDecimal()`. Should we make a copy here or is `otherState` discarded after `combine`?

File: core/trino-main/src/main/java/io/trino/operator/aggregation/DecimalAverageAggregation.java (hunk 8 above)
Comment (martint) [reply core/trino-main/src/main/java/io/trino/operator/aggregation/DecimalAverageAggregation.java:170]: I didn't change the original logic (which I think is confusing and hard to reason about) -- I just restructured the code to avoid repeated calls to get state.

I'll leave that for a future revamp of this class.

File: core/trino-main/src/main/java/io/trino/operator/aggregation/DecimalAverageAggregation.java (hunk 12 above)
Comment (martint) [reply core/trino-main/src/main/java/io/trino/operator/aggregation/DecimalAverageAggregation.java:185]: Same as above... I didn't change the original logic. I just restructured the code to avoid repeated calls to get state.

I'll leave that for a future revamp of this class.

Comment (pettyjamesm) [reply core/trino-main/src/main/java/io/trino/operator/aggregation/DecimalAverageAggregation.java:185]: It’s risky and this is one of those cases where using slices that might be mutated it might be pointers into the middle of a very large allocation and therefore retain it (eg: VariableWidthBlock, the serialized block type of this state) are scary.

My understanding of this and the rest of the code is that this is safe because otherState is dead after combined with this state, and since the addInput call creates a new slice and doesn’t borrow from the input block when the current state is null, the second concern is also not a problem here. That said, this is indeed a fragile interaction. Not any more fragile than it was before, but hopefully something that can be improved with the future work.

**Review by sopel39 — APPROVED** [review]
lgtm % https://github.com/trinodb/trino/pull/6591#discussion_r558686781

File: core/trino-main/src/main/java/io/trino/operator/aggregation/DecimalSumAggregation.java [hunk 13]
```diff
@@ -140,38 +140,41 @@ private static InternalAggregationFunction generateAggregation(Type inputType, T
...
-
-        if (state.getLongDecimal() == null) {
+        Slice sum = state.getLongDecimal();
+        if (sum == null) {
             state.setLongDecimal(otherState.getLongDecimal());
         }
         else {
-            accumulateValueInState(otherState.getLongDecimal(), state);
+            long overflow = UnscaledDecimal128Arithmetic.addWithOverflow(sum, otherState.getLongDecimal(), sum);
+            state.addOverflow(overflow);
```
Comment (pettyjamesm) [inline core/trino-main/src/main/java/io/trino/operator/aggregation/DecimalSumAggregation.java:177]: Shouldn't this be: `state.addOverflow(overflow + otherState.getOverflow())`

**Comment by martint** [comment]
> lgtm % #6591 (comment)

For some reason, that shape of code kills all the perf benefits. Probably some escape analysis issue that's getting in the way.
